File: 0535-encode-and-decode-tinyurl/0535-encode-and-decode-tinyurl.py
class Codec:
    
    def __init__(self):
        # hashmap for encoding
        self.encodeMap ={}
        # hashmap for decoding
        self.decodeMap ={}
        ## a base tiny url
        self.base = "http://tinyurl.com/"
        

    def encode(self, longUrl: str) -> str:
        """Encodes a URL to a shortened URL.
        """
        ## first check if the longurl is already in the encodemap or not 
        if longUrl not in self.encodeMap:
            # then convert the long url into short url
            shortUrl  = self.base + str(len(self.encodeMap) + 1)
            # now encodeMap and decodeMap of current state
            self.encodeMap[longUrl] = shortUrl
            self.decodeMap[shortUrl] = longUrl
            
        return self.encodeMap[longUrl]    
        

    def decode(self, shortUrl: str) -> str:
        """Decodes a shortened URL to its original URL.
        """
        
        ## here the input is already ecoded through previous steps
        
        return self.decodeMap[shortUrl]
        
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(url))

File: 0535-encode-and-decode-tinyurl/0535-encode-and-decode-tinyurl.cpp
class Solution {
    
private:
    unordered_map<string,string> encodeMap;
    unordered_map<string,string> decodeMap;
    
    string base = "http://tinyurl.com/";
    
public:

    // Encodes a URL to a shortened URL.
    string encode(string longUrl) {
        
        if(encodeMap.find(longUrl) == encodeMap.end()){
            
            string shortUrl = base + to_string(encodeMap.size() + 1);
            encodeMap[longUrl] = shortUrl;
            decodeMap[shortUrl] = longUrl;
        }
        
        return encodeMap[longUrl];
        
    }

    // Decodes a shortened URL to its original URL.
    string decode(string shortUrl) {
        
        return decodeMap[shortUrl];
        
    }
};

/*
int main()
{
Codec codec;
string url = "http://tinyurl.com/";
string encoded = codec.encode(url);
string decoded = codec.decode(encoded);

std::cout << "Original URL: " << url << std::endl;
    std::cout << "Encoded URL: " << encoded << std::endl;
    std::cout << "Decoded URL: " << decoded << std::endl;

    return 0;
}*/

// Your Solution object will be instantiated and called as such:
// Solution solution;
// solution.decode(solution.encode(url));